### PR TITLE
Move tests to use azure pipeline credentials 

### PR DESCRIPTION
**Pull request by gearama**
Run all pipelines for the following steps.
- [X] 1. update GetTestCredentials method to use chained creds with pipeline first and default second
image

- [X] 2. update the creds with the proper ENVs for pipeline auth and UseFederatedAuth in ci.yml -- except evenhubs and attestation -- ran all ci and ran live pieplines except attestation and eventhubs
image

- [X] 3 remove the UseFederatedAuth from the individual cy.ymls and move it to the central one, set the storage one to false if needed. 
No idea where the change should go , i asked but got no answer 
 i found the definition in ENG folder in 3 files :
  - live.tests.yml, 
  - archetype-sdk-client.yml 
  - archetype-sdk-tests.yml
 
  Some are in common code , thus the change would have to come down form the ENG repo through a merge into our repo 

- [X] 4 disable KV samples temporarily until we get a resolution to the auth dilema
# Pull Request Checklist

Please leverage this checklist as a reminder to address commonly occurring feedback when submitting a pull request to make sure your PR can be reviewed quickly:

See the detailed list in the [contributing guide](https://github.com/Azure/azure-sdk-for-cpp/blob/main/CONTRIBUTING.md#pull-requests).

- [ ] [C++ Guidelines](https://azure.github.io/azure-sdk/cpp_introduction.html)
- [ ] Doxygen docs
- [ ] Unit tests
- [ ] No unwanted commits/changes
- [ ] Descriptive title/description
  - [ ] PR is single purpose
  - [ ] Related issue listed
- [ ] Comments in source
- [ ] No typos
- [ ] Update changelog
- [ ] Not work-in-progress
- [ ] External references or docs updated
- [ ] Self review of PR done
- [ ] Any breaking changes?

### CODE REVIEW

File: sdk/core/azure-core-test/inc/azure/core/test/test_base.hpp
Comment (gearama) [inline sdk/core/azure-core-test/inc/azure/core/test/test_base.hpp:253]: we cannot use chained token creds as those envs are defined only in the pipeline

File: sdk/storage/ci.yml
Comment (Jinming-Hu) [inline sdk/storage/ci.yml:36]: what does this do?

**Review by Jinming-Hu — APPROVED** [review]
Changes to storage LGTM

File: sdk/storage/ci.yml
Comment (RickWinter) [reply sdk/storage/ci.yml:36]: This switch tells the pipeline if it should use the FederatedAuth (azurepipelinecrdential). Right now storage is NOT ready to use that model so storage is being opted out.

File: samples/helpers/get-env/CMakeLists.txt
Comment (RickWinter) [inline samples/helpers/get-env/CMakeLists.txt:16]: Order them alphabetically

File: samples/helpers/get-env/inc/pipeline_auth_helper.hpp
Comment (RickWinter) [inline samples/helpers/get-env/inc/pipeline_auth_helper.hpp:36]:
```suggestion
};

```

File: samples/helpers/get-env/inc/pipeline_auth_helper.hpp
Comment (RickWinter) [inline samples/helpers/get-env/inc/pipeline_auth_helper.hpp:31]: Does this need to be a catch all? Can we catch a more specific exception.

File: sdk/attestation/ci.yml
Comment (RickWinter) [inline sdk/attestation/ci.yml:35]: Is this necessary? The default for all C++ pipelines is true

File: sdk/core/azure-core-test/inc/azure/core/test/test_base.hpp
Comment (RickWinter) [reply sdk/core/azure-core-test/inc/azure/core/test/test_base.hpp:253]: Is the bug because of GetEnv("") throwing or because we pass null strings to the constructor? Either way, AzurePipelineCredential is broken if it can't be used in the ChainedTokenCredential. Please open a bug for this.

Add a comment here that this is a workaround due to that bug. We should remove the workaround and use a ChainedTokenCredential.

File: sdk/core/perf/inc/azure/perf/base_test.hpp
Comment (RickWinter) [inline sdk/core/perf/inc/azure/perf/base_test.hpp:103]:
```suggestion
     * @note If AZURE_TENANT_ID or AZURE_CLIENT_ID are not available in the
```

File: sdk/core/perf/src/base_test.cpp
Comment (RickWinter) [inline sdk/core/perf/src/base_test.cpp:290]: This should be a chained token credential. See comment above.

File: sdk/eventhubs/ci.yml
Comment (RickWinter) [inline sdk/eventhubs/ci.yml:35]: Is this necessary since the default is true?

File: sdk/identity/azure-identity/samples/CMakeLists.txt
Comment (RickWinter) [inline sdk/identity/azure-identity/samples/CMakeLists.txt:34]: Why are all the samples commented out?

File: sdk/identity/azure-identity/test/ut/azure_pipelines_credential_test.cpp
Comment (RickWinter) [inline sdk/identity/azure-identity/test/ut/azure_pipelines_credential_test.cpp:647]: Why is this commented out?

File: sdk/identity/azure-identity/test/ut/token_credential_test.cpp
Comment (RickWinter) [inline sdk/identity/azure-identity/test/ut/token_credential_test.cpp:58]: Call GetEnv once to get the value and then compare that value.

string envVar = GetEnv("AZURE_TEST_MODE");
if (envVar == "LIVE" || envVar == "live)

Comment (RickWinter) [reply sdk/identity/azure-identity/test/ut/token_credential_test.cpp:58]: Or even better would be to put this logic into a test helper function then do `if(IsLiveTestMode())`

File: sdk/identity/ci.yml
Comment (RickWinter) [inline sdk/identity/ci.yml:33]: Is this necessary if default is true?

File: sdk/keyvault/ci.yml
Comment (RickWinter) [inline sdk/keyvault/ci.yml:35]: Default?

File: sdk/tables/ci.yml
Comment (RickWinter) [inline sdk/tables/ci.yml:33]: This is default so is it necessary?

File: samples/helpers/get-env/CMakeLists.txt
Comment (gearama) [reply samples/helpers/get-env/CMakeLists.txt:16]: they are abcgef..G..hijklmno....P ....

File: samples/helpers/get-env/inc/pipeline_auth_helper.hpp
Comment (gearama) [reply samples/helpers/get-env/inc/pipeline_auth_helper.hpp:31]: i don't know which one , it's either from reading the env, or the constructor fails , or the auth failes

File: sdk/core/perf/src/base_test.cpp
Comment (gearama) [reply sdk/core/perf/src/base_test.cpp:290]: can't be , azure pipeline creds fails due to lack of envs and their content

File: sdk/eventhubs/ci.yml
Comment (gearama) [reply sdk/eventhubs/ci.yml:35]: yes, see previous comments and PR message

File: sdk/identity/azure-identity/samples/CMakeLists.txt
Comment (gearama) [reply sdk/identity/azure-identity/samples/CMakeLists.txt:34]: as i mentioned in the PR due to issues with creds in live they fail to auth , the sample fails , the live tests fail

File: sdk/identity/azure-identity/test/ut/azure_pipelines_credential_test.cpp
Comment (gearama) [reply sdk/identity/azure-identity/test/ut/azure_pipelines_credential_test.cpp:647]: somehow DISABLE does not disample this test

File: sdk/identity/ci.yml
Comment (gearama) [reply sdk/identity/ci.yml:33]: yes

File: sdk/keyvault/ci.yml
Comment (gearama) [reply sdk/keyvault/ci.yml:35]: nope, can't do, can't figure out where

File: sdk/tables/ci.yml
Comment (gearama) [reply sdk/tables/ci.yml:33]: yes

File: sdk/core/azure-core-test/inc/azure/core/test/test_base.hpp
Comment (gearama) [reply sdk/core/azure-core-test/inc/azure/core/test/test_base.hpp:253]: https://github.com/Azure/azure-sdk-for-cpp/issues/5769

**Review by antkmsft — CHANGES_REQUESTED** [review]
Is there another way? The samples should to illustrate how the users are supposed to use the SDK, i.e. start with DefaultAzureCredential, seeing "GetSampleCredential" there defeats some pupopse of a sample.

**Comment by weshaggard** [comment]
> Is there another way? The samples should to illustrate how the users are supposed to use the SDK, i.e. start with DefaultAzureCredential, seeing "GetSampleCredential" there defeats some pupopse of a sample.

JS team used this workaround https://github.com/Azure/azure-sdk-for-js/pull/30132 to still use DAC but avoid the managed identity.

File: sdk/identity/azure-identity/src/default_azure_credential.cpp
Comment (ahsonkhan) [inline sdk/identity/azure-identity/src/default_azure_credential.cpp:46]: Please revert this change along with the changes to `AzurePipelinesCredential` public surface area. This credential is **by design** not part of the DAC across all languages, and it isn't intended to be.

If you could summarize the problem you are trying to solve here that required such a change, I can try and suggest alternative solutions. I recommend seeing what other languages are doing to get inspiration.

File: sdk/identity/azure-identity/inc/azure/identity/azure_pipelines_credential.hpp
Comment (ahsonkhan) [inline sdk/identity/azure-identity/inc/azure/identity/azure_pipelines_credential.hpp:101]: We cannot add such an API with these default env vars.

Sharing the gist some the technical details/design context for such a decision: None of the languages do this today, and the env vars may not be stable. We're waiting until a cleaner solution is available as stable by the service, before exposing such conveniences.

**Review by ahsonkhan — CHANGES_REQUESTED** [review]
I only took a quick review of the Identity related changes, and will follow-up once I am back.

Please revert the changes to `AzurePipelinesCredential` and its public surface area. Also, remove it from the DAC.

File: sdk/identity/azure-identity/inc/azure/identity/azure_pipelines_credential.hpp
Comment (RickWinter) [reply sdk/identity/azure-identity/inc/azure/identity/azure_pipelines_credential.hpp:101]: I believe this is a mistake by the Identity working group. I spoke with Christopher in length about this. The belief that a user passes them verse after reading from the Env rather than the credential reading from the Env is really passing the buck and provides no security benefit.

File: sdk/identity/azure-identity/src/default_azure_credential.cpp
Comment (RickWinter) [reply sdk/identity/azure-identity/src/default_azure_credential.cpp:46]: The credential should be part of the DAC. Its the preferred credential for Azure pipelines as part of SFI. Not having it in the DAC really brings to light that no one should use DAC.

Comment (ahsonkhan) [reply sdk/identity/azure-identity/src/default_azure_credential.cpp:46]: Did this change occur recently, across all languages that implemented the AzurePipelinesCredential? Please share more details of this design discussion and conclusion, offline. If there is a documented doc that describes the motivation behind this change, let me know as I would love to read up on that.

Comment (ahsonkhan) [reply sdk/identity/azure-identity/src/default_azure_credential.cpp]: Here's the design gist that the languages followed:
https://gist.github.com/KarishmaGhiya/81ee6265ab6e9109d3bf510678878b34/

**Comment by gearama** [comment]
> I only did a quick review of the Identity related changes, and will follow-up once I am back.
> 
> Please revert the changes to `AzurePipelinesCredential` and its public surface area. Also, remove it from the DAC. None of the languages have it there, by design.

Done

**Review by RickWinter — APPROVED** [review]

File: sdk/identity/azure-identity/samples/CMakeLists.txt
Comment (antkmsft) [inline sdk/identity/azure-identity/samples/CMakeLists.txt:34]: If running these samples fails, let's change the CI to still build them without running.

**Review by antkmsft — COMMENTED** [review]
I thought the plan was to add `AzurePipelinesCredential` into the list of sources for `DefaultAzureCredential`?

BTW, how do the rest of samples succeed now, if they are still using DAC?

File: eng/pipelines/templates/jobs/live.tests.yml
Comment (ahsonkhan) [inline eng/pipelines/templates/jobs/live.tests.yml:253]: What's motivating this change to move this as an else?

File: sdk/core/azure-core-test/inc/azure/core/test/test_base.hpp
Comment (ahsonkhan) [inline sdk/core/azure-core-test/inc/azure/core/test/test_base.hpp:317]: nit:
```suggestion
     * @note If AZURE_TENANT_ID or AZURE_CLIENT_ID are not available in the
```

File: sdk/keyvault/azure-security-keyvault-keys/README.md
Comment (ahsonkhan) [inline sdk/keyvault/azure-security-keyvault-keys/README.md:82]: Is this doc change sufficient in explaining to the user what do to without using a client_secret?

File: sdk/storage/README.md
Comment (ahsonkhan) [inline sdk/storage/README.md:39]: Revert change before merging.

File: sdk/keyvault/azure-security-keyvault-secrets/CMakeLists.txt
Comment (ahsonkhan) [inline sdk/keyvault/azure-security-keyvault-secrets/CMakeLists.txt:120]: Rather than changing the CMakeLists.txt, can we update the ci leg that's passing in `BUILD_SAMPLES=ON` in the yml, and we can revert that after?

File: sdk/identity/azure-identity/samples/CMakeLists.txt
Comment (ahsonkhan) [inline sdk/identity/azure-identity/samples/CMakeLists.txt:42]: Please explain why we are commenting all the samples out.

File: eng/pipelines/templates/stages/archetype-sdk-client.yml
Comment (ahsonkhan) [inline eng/pipelines/templates/stages/archetype-sdk-client.yml:66]: What's requiring this change? I don't understand the comment. What is a "config for Preview"?

**Review by antkmsft — APPROVED** [review]
Approving to unbock the CI

File: sdk/keyvault/azure-security-keyvault-keys/README.md
Comment (gearama) [reply sdk/keyvault/azure-security-keyvault-keys/README.md:82]: should suffice , i removed trhe ENV form there

File: sdk/identity/azure-identity/samples/CMakeLists.txt
Comment (gearama) [reply sdk/identity/azure-identity/samples/CMakeLists.txt:34]: whish i could , no idea how without disabling the run samples steps for everyone

File: eng/pipelines/templates/jobs/live.tests.yml
Comment (gearama) [reply eng/pipelines/templates/jobs/live.tests.yml:253]: the if statement is IF use AzrurePipelineCredentials define these ENVs , ti was upside down

File: sdk/keyvault/azure-security-keyvault-secrets/CMakeLists.txt
Comment (gearama) [reply sdk/keyvault/azure-security-keyvault-secrets/CMakeLists.txt:120]: i'd rather avoid messing with yml files, also i don't want to disable globally for everyone

File: sdk/storage/README.md
Comment (gearama) [reply sdk/storage/README.md:39]: done

File: eng/pipelines/templates/jobs/live.tests.yml
Comment (ahsonkhan) [reply eng/pipelines/templates/jobs/live.tests.yml:253]: @weshaggard since Daniel is OOF, can you please help review the changes in this `live.tests.yml` file. I think it could benefit from an extra set of eyes to make sure it is as expected :)

File: eng/pipelines/templates/stages/archetype-sdk-client.yml
Comment (gearama) [reply eng/pipelines/templates/stages/archetype-sdk-client.yml:66]: that was introduced by Daniel in my PR based on the comment this is needed for the Preveiw cloud , whatever thast is

File: eng/pipelines/templates/jobs/live.tests.yml
Comment (gearama) [reply eng/pipelines/templates/jobs/live.tests.yml:253]: it was very easy to check , before my change the build the live tests were failing, after it was working 
and if you read the line right above it you see the if statement

Comment (weshaggard) [reply eng/pipelines/templates/jobs/live.tests.yml:253]: It looks correct assuming you want to continue to run the samples via default credentials instead of with pipeline credentials.

Comment (gearama) [reply eng/pipelines/templates/jobs/live.tests.yml:253]: Daniel was aware of the change, i talked to him before making it

Comment (gearama) [reply eng/pipelines/templates/jobs/live.tests.yml:253]: [7/3 10:01 AM] George Arama
from what i see in the live.tests.yml the 
- ${{ if parameters.UseFederatedAuth }}:
.....
-${{ else }}:
..
$env:AZURESUBSCRIPTION_CLIENT_ID = $account.Id

$env:AZURESUBSCRIPTION_TENANT_ID = $account.Tenants
 
...
 
the variables get defined when not using federated auth on the False branch 
 
[7/3 10:24 AM] George Arama
in the live tests yaml the if/else branches for the run  samples are flipped , i'm testing my fix for it 
 
[7/3 11:13 AM] George Arama
looks like ti's working now as expected

File: sdk/identity/azure-identity/test/ut/azure_pipelines_credential_test.cpp
Comment (ahsonkhan) [inline sdk/identity/azure-identity/test/ut/azure_pipelines_credential_test.cpp:647]: Rather than commenting out the test, please disable it, by renaming the test like this (prefix with `DISABLED_`):
https://github.com/Azure/azure-sdk-for-cpp/pull/5785

> DISABLED_InvalidSystemAccessToken_LIVEONLY_

See:
https://dev.azure.com/azure-sdk/internal/_build/results?buildId=3954813&view=logs&j=4548f7fc-05cc-57aa-a896-3f4bc48e7aad&t=a90428ec-b6a1-582b-bdc6-58150785e27b

> 38 - azure-identity.AzurePipelinesCredential.InvalidSystemAccessToken_LIVEONLY_ (Disabled)

File: sdk/identity/azure-identity/samples/CMakeLists.txt
Comment (ahsonkhan) [reply sdk/identity/azure-identity/samples/CMakeLists.txt:42]: I am not sure I understand the issue.

It's reasonable for us to disable building/running the samples if we know the problem and have a plan to update it (even in a separate, follow-up PR).

Can you share the error you were seeing or the link to the failing pipeline that required us to comment these out?

Comment (ahsonkhan) [reply sdk/identity/azure-identity/samples/CMakeLists.txt:42]: Looking at the pipelines that build and run Identity samples, they seem to pass from this PR, without commenting them out:
https://github.com/Azure/azure-sdk-for-cpp/pull/5785

https://dev.azure.com/azure-sdk/internal/_build/results?buildId=3955376&view=logs&j=ea109acd-b7d3-5ded-4f2b-e0ac8b0c5651&t=88917f58-d2f9-5f14-5936-80f13f049405

https://dev.azure.com/azure-sdk/internal/_build/results?buildId=3955376&view=logs&j=fad7e5d9-ee9d-5455-29cf-12d77fc3497f&t=e819b337-613f-5b6a-fedf-c6b8af28994f

File: sdk/identity/azure-identity/test/ut/azure_pipelines_credential_test.cpp
Comment (ahsonkhan) [reply sdk/identity/azure-identity/test/ut/azure_pipelines_credential_test.cpp:647]: `DISABLED_` needs to be a **prefix** of the test name, not suffix.

Comment (ahsonkhan) [reply sdk/identity/azure-identity/test/ut/azure_pipelines_credential_test.cpp:647]: Please pull from main which contains the changes from https://github.com/Azure/azure-sdk-for-cpp/pull/5785.

That way, you don't have to modify the identity test files in this PR, at all.

**Review by ahsonkhan — APPROVED** [review]
We certainly need to revisit the decision to comment out the identity samples and see how we can turn them back on.
I don't fully understand the changes to `live.tests.yml` that flip the if/else and the implications of doing so, but it seems reasonable from a quick glance.

Approving to unblock the change since we can revisit this discussion in a follow-up.

File: sdk/identity/azure-identity/samples/CMakeLists.txt
Comment (antkmsft) [reply sdk/identity/azure-identity/samples/CMakeLists.txt:42]: resolving this conversation to merge the CI

File: eng/pipelines/templates/jobs/live.tests.yml
Comment (antkmsft) [reply eng/pipelines/templates/jobs/live.tests.yml:253]: resolving this conversation to merge the CI

**Pull request merged**
